### ISSUE TITLE
f16 with WGPU

### ISSUE THREAD

**Issue by Gadersd** [tracel-ai/burn#597]
### Feature description

It would be great if burn-wgpu supported f16. Is there a timeline for this?

### Feature motivation

Large models such as Stable Diffusion exceed wgpu's maximum buffer size when using f32. f16 support would enable some such models to be run with the wgpu backend.

**Comment by nathanielsimard** [comment]
Linked to https://github.com/gfx-rs/naga/issues/1884

**Comment by Gadersd** [comment]
Does anyone know if the limited buffer size in wgpu will be alleviated eventually? Even if f16 gets supported the buffer size limits will still be a barrier to running large models.

**Comment by nathanielsimard** [comment]
You can manually override the limits when selecting the device : https://github.com/burn-rs/burn/blob/ed255c5561b85876cf02cbc4d48f35e1f0d29ac0/burn-wgpu/src/context/base.rs#L228

The limits are low for compatibility reasons I think, but I can increase `max_storage_buffer_binding_size` on my RTX 3070 to `usize::pow(8, 10)` and load bigger tensors. I think we should come up with a way to change the limits for specific devices, probably with a config file or env variables (or both).

**Comment by Gadersd** [comment]
Would it be reasonable to use the pub fn [limits](https://docs.rs/wgpu/latest/wgpu/struct.Adapter.html#method.limits)(&self) -> [Limits](https://docs.rs/wgpu/latest/wgpu/struct.Limits.html) function on the adapter to get the best limits that the adapter offers instead of relying on defaults? I think this would resolve the issue.

**Comment by nathanielsimard** [comment]
@Gadersd PR done https://github.com/burn-rs/burn/pull/601 Let me know if it helps in running your models.

**Comment by Gadersd** [comment]
Success! Stable Diffusion now runs on wgpu.

**Comment by Gadersd** [comment]
My bad, I accidentally tested with tch.

**Comment by nathanielsimard** [comment]
It may happen when you run out of memory. You can try to lower the `MAX_TASKS` to 1 to reduce memory usage:

https://github.com/burn-rs/burn/blob/9361193b5d62065807fdb6721e95dca8bcf8bf74/burn-wgpu/src/context/server.rs#L65

It might increase a bit the computing time, but it's probably negligible for a big model.

**Comment by Gadersd** [comment]
Setting `MAX_TASKS` to 1 enabled inference to work, but it was very slow compared to the tch run, ~5 minutes vs ~15 seconds. Perhaps the value should be settable by the user when the default isn't viable?

**Comment by nathanielsimard** [comment]
Yes we could do that for now. There is an issue to optimize the memory strategy: https://github.com/burn-rs/burn/issues/582.

**Comment by nathanielsimard** [comment]
@Gadersd I added a way to configure MAX_TASKS: https://github.com/burn-rs/burn/pull/603.

**Comment by antimora** [comment]
There is a tweet saying "float16 in webGPU finally works now"

https://twitter.com/nisten/status/1698796718840598850

Worth looking into this and see if we need to update anything.

**Comment by antimora** [comment]
https://github.com/gfx-rs/wgpu/issues/4384

**Comment by antimora** [comment]
CC: @wingertge @ArthurBrussee

**Comment by antimora** [comment]
Retested on main: f16 matmul works on both `wgpu` (WGSL) and `metal`. Closing.
